File: bigfactors2.py
import PIL, math
import numpy as np

#Problem 12 Highly divisible triangular number
#finds the first number with over 500 factors

def countfactors(num):
	factors = 0
	### we only need to know about the FIRST HALF of factors,
	##one factor implies a second
	root = int(math.ceil(math.sqrt(num)))
	divs = range(1, root)
	for d in divs:
		if(num%d == 0):
			factors += 2
			
	#Correction if the number is a perfect square
	if (root * root == num):
		factors-=1
	return factors


####   MAIN #####
i=1
k=1
j = 0
while(k < 500):
	j += i
	k=countfactors(j)
	print str(j) + " has " + str(k) + " factors"
	i += 1

File: bigfactors2.c
#include <stdlib.h>
#include <stdio.h>
#include <math.h>

int countFactors(int num){
	int factors = 0;
    // Check only up until the square root of the number
 	int root = (int) ceil(sqrt(num));
    //printf("%d\n", root);
    for(int i = 2; i < root; i++){
        if(num % i == 0)
            factors+=2;
    }
    // Correction for perfect square
    if(root * root == num)
        factors -= 1;
    return factors;
}

int main(){
	int i = 1;
	int k = 1;
	int j = 0;
	while(k < 500){
		j += i;
		k = countFactors(j);
		i += 1;
	}
	printf("%d has over 500 factors. Neat!\n", j);
	return 0;
}
